### ISSUE TITLE
Add Sizzle.noConflict()

### ISSUE THREAD

**Issue by andriytkachiv** [jquery/sizzle#352]
Could you use **this** instead of **window** in self executing anonymous wrapper for whole code?
I am developing third party widget and need to attach Sizzle to my custom object, so it won't pollute global scope. I wrap all my code with this ```;var myObject = {}; (function(){ /*here comes all widget's code*/ }).call(myObject);'``` and in my case  ```this != window```

**Comment by mzgol** [comment]
Sizzle needs the real `window` object (or its very good implementation like jsdom); why would you want to use anything else?

Or are you just talking about the object to which the `Sizzle` instance is attached? In that case, I can imagine some people may be wrapping Sizzle in an IIFE with a non-window `this` and still expect `Sizzle` to be defined under `window` so it seems pretty arbitrary. You can always do:
```js
// Sizzle code
var Sizzle = window.Sizzle;
delete window.Sizzle;
```

**Comment by andriytkachiv** [comment]
@mzgol Yes, I talked only about attached object. 
Your solution is not very clear, but still better than forking ```Sizzle``` repository and changing few lines :) 
Thanks!

**Comment by andriytkachiv** [comment]
However, not. Because if user uses Sizzle on his website and included widget will also use it in this way, than ```delete window.Sizzle;``` will cause problems. 
Also user's version of ```Sizzle``` could be different from widget's, and it will also cause problems.

It's quite rare case, but still possible.

**Comment by mzgol** [comment]
If you have to have `window.Sizzle` defined then I don't understand the problem... What exactly do you want to achieve?

**Comment by andriytkachiv** [comment]
I don't need ```window.Sizzle```, but I also couldn't delete it, if it exists.

I have widget which user include in html code on his website. Widget loads javascript file which is injected inline, not as iframe. I use them some small and efficient libraries, like domready, Sizzle, etc.
As widget comes inline it couldn't pollute global scope, because user could use any of this library or global names. To achieve this I wrap all code as I mentioned above and everything is assigned to global object. 

If user do not use Sizzle on his website there will no problem, but if he will, what will happend in that case?

**Comment by mzgol** [comment]
I see; you need to include Sizzle in the widget but you don't want to break a potential another copy already existing on the site.

It seems you'd need a `Sizzle.noConflict()` method similar to [the jQuery one](https://github.com/jquery/jquery/blob/a644101ed04d0beacea864ce805e0c4f86ba1cd1/src/exports/global.js#L8-L18) that you'd be able to use right after including Sizzle.

I'll leave it to @gibson042 to decide but it seems a sensible idea. Thanks for the report.

**Comment by andriytkachiv** [comment]
Yes, that's exactly what I want!
Thank you very much!

**Comment by gibson042** [comment]
Thus far, Sizzle has been billed as a library building tool, with the expectation that consumers would replace the [`EXPOSE` block](https://github.com/jquery/sizzle/blob/3b1abbab1fb0c8080f36ad889d37704c7be13641/src/sizzle.js#L2131-L2140) for their own purposes [as jQuery does](https://github.com/jquery/jquery/blob/63a577aa0bcb439c1730c3825407d86c128b17be/build/tasks/build.js#L59-L60). However, the default state of that block is intended to be useful on its own, and I'm comfortable with adding such functionality inside it.

**Comment by mzgol** [comment]
@andriytkachiv Would you like to take a stab at a PR for that? If not, we'll do it one day.

**Comment by andriytkachiv** [comment]
@mzgol  It will be nice if I could do PR :)

**Comment by mzgol** [comment]
Go ahead then. :)

**Comment by andriytkachiv** [comment]
@mzgol Okay :) I gues, you decided to implement solution "add exports variable to self-executing anonymous function, and expose section will bind to this variable", am I right?

**Comment by mzgol** [comment]
@andriytkachiv You need to write something like [src/exports/global.js in jQuery](https://github.com/jquery/jquery/blob/9d820fbde6d89bc7a06e2704be61cf6c0b4d6e3c/src/exports/global.js), just without the `noGlobal` part and put it inside the [Sizzle EXPOSE block](https://github.com/jquery/sizzle/blob/3b1abbab1fb0c8080f36ad889d37704c7be13641/src/sizzle.js#L2131-L2140) (@gibson042 linked to it before). This will then not influence jQuery in any way as it strips that out but when used standalone there will be a `Sizzle.noConflict()` method.

**Comment by andriytkachiv** [comment]
@mzgol  Thanks for tip :) Hope, tomorrow will send PR!

**Comment by gibson042** [comment]
@andriytkachiv Are you still planning to submit a PR?

**Comment by andriytkachiv** [comment]
@gibson042  I am very sorry, unfortunately I have a lot to do and completely forgot about it. 
At this moment I use jQuery instead and don't need this functionality. But this could be great for other users, so will be very thankful if you could add it, because It will takes me some time to integrate with your project, install all this staff.
Thanks.

**Comment by SaptakS** [comment]
I would like to solve this issue. Can you tell me how I can recreate the problem? What I have understood is if a single page has 2 different versions of sizzle, then there can be a conflict, just like in case of jQuery. So, we need to implement something similar to jQuery.noConflict if I am not wrong. Please guide me so that I can submit a PR.

**Comment by mzgol** [comment]
@SaptakS Could you ask more specific questions? I've already written what needs to be done in https://github.com/jquery/sizzle/issues/352#issuecomment-134397943, is there something not clear there?

**Comment by SaptakS** [comment]
I was clarifying whether I have understood it correctly. And wanted to know how to recreate the issue.

**Comment by gibson042** [comment]
The test will be similar to mixed_sort:
* An HTML document in `test/data/` that loads Sizzle, stores a reference to it, loads Sizzle again and stores another reference, then passes assertion results to the parent window callback:
  * Before `Sizzle.noConflict()`:
    * `Sizzle !== Sizzle1`
    * `Sizzle( selector )` and `Sizzle1( selector )` yield identical (and correct) results
  * After `Sizzle.noConflict()`:
    * `Sizzle === Sizzle1`
    * `Sizzle( selector )` and `Sizzle2( selector )` yield identical (and correct) results
* Updates to `test/karma/karma.conf.js` for the same treatment as `test/data/mixed_sort.html`
* A `testIframeWithCallback` in `test/unit/utilities.js`

**Comment by SaptakS** [comment]
OK. I will start the work asap. Will ask if I face any further problem.

**Comment by SaptakS** [comment]
@gibson042 I am having some problem. I tried this out before writing any noConflict according to what you said http://fpaste.org/285934/44644997/ . So I should be getting "conflict" as output, right? But I get "no conflict". Am I missing something? Please help.

**Comment by mzgol** [comment]
@SaptakS One line before the `if` you assign` Sizzle1 = Sizzle` so they're the same; this has nothing to do with Sizzle in particular, it's just JavaScript. Also, you have a typo in the `else if` branch.

**Comment by SaptakS** [comment]
Ok. Thanks.

**Comment by gibson042** [comment]
Closed by 0561fc35d8fe0680ae928bf3af25d43a7d221a75.
